content/algorithms/vf2_vs_vf2pp/vf2_vs_vf2pp.md

```diff
@@ -0,0 +1,782 @@
+)
+axs[0][1].set_title("p = 0.3")
+axs[0][1].set_ylabel("Execution Time (seconds)")
+axs[0][1].set_xlabel("Number of nodes")
+
+(vf2_line2,) = axs[1][0].plot(
+    x, vf2_all_times[2], label="vf2", linestyle="--", marker="o"
+)
+(vf2pp_line2,) = axs[1][0].plot(
+    x, vf2pp_all_times[2], label="vf2++", linestyle="--", marker="o"
+)
+axs[1][0].set_title("p = 0.5")
+axs[1][0].set_ylabel("Execution Time (seconds)")
+axs[1][0].set_xlabel("Number of nodes")
+
+(vf2_line2,) = axs[1][1].plot(
+    x, vf2_all_times[3], label="vf2", linestyle="--", marker="o"
+)
+(vf2pp_line2,) = axs[1][1].plot(
+    x, vf2pp_all_times[3], label="vf2++", linestyle="--", marker="o"
+)
+axs[1][1].set_title("p = 0.6")
+axs[1][1].set_ylabel("Execution Time (seconds)")
+axs[1][1].set_xlabel("Number of nodes")
+
+axs[0][0].legend(handles=[vf2_line1, vf2pp_line1])
+axs[0][1].legend(handles=[vf2_line2, vf2pp_line2])
+axs[1][0].legend(handles=[vf2_line1, vf2pp_line1])
+axs[1][1].legend(handles=[vf2_line2, vf2pp_line2])
+
+fig.tight_layout(pad=2.0)
+plt.show()
+```
+
+In both graphs and directed graphs, the experimental results are quite similar. VF2++ continues to outperform VF2. The growth of VF2's execution time increases exponentially as the number of nodes increases, while VF2++ has a linear behavior. Additionally, it's noteworthy that for all values of $p$, the curves of VF2 and VF2++ are very similar, with the only difference being the increase in execution time as the graphs become denser.
+
++++
+
+## Create graphs from a degree secuence
+
+To test on negative instances, in other words, graphs that are not isomorphic we need to create graphs with the same degree sequence. For that, we can use *nx.configuration_model(degree_sequence)* that creates a random MultiGraph with the given degree sequence. The degree sequence must have an even sum. Then we can just convert the resulting MultiGraph into a Graph. Let's see in an example how can we experiment with this:
+
+```{code-cell}
+# Example
+import random
+
+random.seed(11)  # Set a random seed
+d = random.choices(
+    list(range(0, 5)), k=11
+)  # Create a list with 11 random numbers in [0, 5]
+
+print("Is the sequence sum even?", np.sum(d) % 2 == 0)
+
+# Create 2 different random graphs with the same degree secuence
+g1 = nx.Graph(nx.configuration_model(d, seed=10))
+g1.remove_edges_from(nx.selfloop_edges(g1))
+
+g2 = nx.Graph(nx.configuration_model(d, seed=11))
+g2.remove_edges_from(nx.selfloop_edges(g2))
+
+print(g1)
+print("Are isomorphic? ", nx.is_isomorphic(g1, g2))
+```
+
+To do this same experiment on DiGraphs, we need to define an in-degree sequence and an out-degree sequence.  
+We will use the function *nx.directed_configuration_model(in_degree_sequence, out_degree_sequence, create_using, seed)*. Both degree sequences must have the same sum. Again this function returns a MultiDiGraph so we will have to convert it into a DiGraph. Let's do this in an example:
```

Comment:
I'm surprised that the resulting degree sequences are the same for the two graphs. The multiedges are not likely to be the same -- and the selfloop edges are also likely to be a different number of edges with the two seeds. So when you convert to a non-multigraph, and remove the selfloops, you change the degree sequences.  And it seems like you would be changing them in different ways. Have you checked that the two graphs have the same degree sequence?  If they don't, the function will correctly return `True`, but it won't actually run any of the algorithm to find this out. I think it doesn't even create the data structures that would cost time for overhead in VF2++, but I'm not sure about that.